core/textile/sync/queue.go

```diff
@@ -0,0 +1,132 @@
+package sync
+
+import (
+	"container/list"
+	"encoding/json"
+	"fmt"
+
+	"github.com/FleekHQ/space-daemon/log"
+)
+
+const QueueStoreKey = "TextileSyncTaskQueue"
+
+type marshalledQueue struct {
+	queueAsSlice     []*Task          `json:"queueAsSlice"`
+	fileQueueAsSlice []*Task          `json:"fileQueueAsSlice"`
+	hashMap          map[string]*Task `json:"queueHashMap"`
+}
+
+func (s *synchronizer) enqueueTask(task *Task, queue *list.List) {
+	queue.PushBack(task)
+	s.queueHashMap[task.ID] = task
+}
+
+func (s *synchronizer) dequeueTask(queue *list.List) *Task {
+	queueItem := queue.Front()
+	s.taskQueue.Remove(queueItem)
+
+	task := queueItem.Value.(*Task)
+	delete(s.queueHashMap, task.ID)
+
+	return task
+}
+
+func (s *synchronizer) storeQueue() error {
+	// Store main queue
+	queueAsSlice := []*Task{}
+	currEl := s.taskQueue.Front()
+
+	for currEl != nil {
+		queueAsSlice = append(queueAsSlice, currEl.Value.(*Task))
+		currEl = currEl.Next()
+	}
+
+	// Store file pinning queue
+	fileQueueAsSlice := []*Task{}
+	currEl = s.filePinningQueue.Front()
+
+	for currEl != nil {
+		fileQueueAsSlice = append(fileQueueAsSlice, currEl.Value.(*Task))
+		currEl = currEl.Next()
+	}
+
+	objToMarshal := &marshalledQueue{
+		queueAsSlice:     queueAsSlice,
+		fileQueueAsSlice: fileQueueAsSlice,
+		hashMap:          s.queueHashMap,
```

Comment:
@dmerrill6 

Does it duplicate the same `Task` as in the slices?

What is the plan for `hashMap` except for a (unused) `isTaskEnqueued` check?

Reply:
> Does it duplicate the same Task as in the slices?
Not sure what you mean by this

What is the plan for hashMap except for a (unused) isTaskEnqueued check?
I did it initially in case I wanted to access an element in the queue directly, so that I could get access in O(1) instead of finding it in the linked list, but ended up not using it. However after implementing your suggestion about task IDs, I'm now using it when enqueuing a new task. First it will see if the task exists in the queue and avoid adding it if it's already there.

Reply:
Stopped storing the hash map in f4d3cc6